### ISSUE TITLE
Specify quantization type for tflite exports (per-channel / per-tensor)

### ISSUE THREAD

**Issue by kfrohwergmh** [ultralytics/ultralytics#24747]
### Search before asking

- [x] I have searched the Ultralytics [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar feature requests.

### Description

I would like to be able to customize the quantization type when exporting to TFLite. 
Currently, “per-channel” is selected by default, as this is the default behavior in onnx2tf.convert. 
However, my hardware requires “per-tensor” quantization.
I suggest that this default behavior remain in place, but that the option to customize the quantization type be added.
To do this, there would need to be a parameter that can be specified during export.

### Use case

Expected behavior:

```python
from ultralytics import YOLO
import argparse

def get_commandline_arguments():
    parser = argparse.ArgumentParser(description="Export YOLO model to TFLite format")
    parser.add_argument("--model", type=str, required=True, help="Path to the trained YOLO model")
    parser.add_argument("--data", type=str, required=True, help="Path to the calibration data YAML file")
    return parser.parse_args()

args = get_commandline_arguments()

data_cal_path = args.data  # path to the calibration data YAML file
model = YOLO(args.model)

# for per-channel quantization
model.export(format="tflite", int8=True, data=data_cal_path, end2end=False, imgsz=640)
# for per-tensor quantization
model.export(format="tflite", int8=True, quant_type="per-tensor", data=data_cal_path, end2end=False, imgsz=640)
```

### Additional

Here's the difference between the two models after I implemented it:

<img width="1674" height="345" alt="Image" src="https://github.com/user-attachments/assets/c54f9a5b-1aad-4630-9f3c-66d83df853e5" />

On my hardware this results in an huge decrease of inferencetime.

|                     | per-channel | per-tensor |
|:-------------------:|:-----------:|:------------:|
| Mean Inference Time | 1764 ms     | 42 ms      |

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @kfrohwergmh, thank you for your interest in Ultralytics 🚀 and for the detailed feature request around TFLite quantization export options. This is an automated response to help guide the issue process, and an Ultralytics engineer will also assist you soon 😊

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python) and [CLI](https://docs.ultralytics.com/usage/cli) usage examples and where many of the most common questions may already be answered.

Since this appears to be a feature request rather than a bug report, thank you as well for including a clear use case, example API proposal, and performance comparison. If you move forward with a PR, please include as much implementation detail and validation context as possible so the team can review it efficiently.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes) and [Tasks](https://docs.ultralytics.com/tasks) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
`per-tensor` quantization results in broken INT8 model from our experience, so it's not available as an option

**Comment by kfrohwergmh** [comment]
Oh, okay. 
The one i had produce so far is up and running. I evaluated it and there is no noticable difference between the pt model and the per-tensor quantized int8 model.
It should be noted, however, that this is a classification model in which I replaced the YOLO backbone with a ResNet backbone.

I tried this with a detection model as well. However, my accuracy drops significantly in that case. I attributed the difference to the quantized detection head and the fact that I only want to detect a single class of objects in the image. Also, my dataset comes from a stationary camera, so there isn’t a wide variety of examples in the calibration dataset.
To put it briefly, is this also a sign of a broken model? In my view, there are plenty of other reasons (as statet above) why the model is losing accuracy.
It works fine (mAP50 ~ 0.7/0.8) and detects objects in the image with an acceptable inferencetime. 

So my final question is, what are the symptoms of a broken INT8 model or what should I look out for out of your experience?

**Comment by glenn-jocher** [comment]
If the INT8 export is stable and its validation metrics stay reasonably close to the PT baseline on a held-out set, I would not call it broken; the usual red flags are class collapse, boxes snapping to fixed locations/sizes, confidences saturating near 0, predictions disappearing or exploding on similar images, or a large unexplained drop after `YOLO("model_int8.tflite").val(data=...)`. For detection, `per-tensor` tends to hurt the head/box decoding much more than classification, so your results are consistent with what we’ve seen, and the best check is a side-by-side PT vs TFLite validation using representative calibration data as described in the [TFLite export docs](https://docs.ultralytics.com/integrations/tflite).

**Comment by kfrohwergmh** [comment]
Hmm, I see. Do you know of any workarounds or settings I can use to prevent the model from getting corrupted, for example, by avoiding certain operators? Or in other words, do you know why the model brokes when using per-tensor quantization?

**Comment by Y-T-G** [comment]
Which TFLite file are you testing? The dynamic quantized model works fine. The static quantized one is broken

**Comment by kfrohwergmh** [comment]
I did not set dynamic=True so I think i am using the static quantized one.

**Comment by glenn-jocher** [comment]
Not necessarily—`dynamic=True` is not what selects the TFLite quantization mode here; with `model.export(format="tflite", int8=True)`, Ultralytics returns `<model>_int8.tflite` with FP32 I/O per the [exporter reference](https://docs.ultralytics.com/reference/engine/exporter), while the fully static/full-integer artifact is a different file inside the `_saved_model` directory, so please check the exact `.tflite` filename you benchmarked.

**Comment by kfrohwergmh** [comment]
ah, i see. I benchmarked the full-integer quantized one `best_full_integer_quant.tflite`. Are there any known workarounds to repair it? It would be nice, because my hardware accelerates with per-tensor quantization and needs a full-integer quantized model.

**Comment by glenn-jocher** [comment]
For detection, not currently—when `best_full_integer_quant.tflite` breaks with `per-tensor`, it is usually quantization error in the detection head/box decode rather than one operator we can safely disable, so there is no reliable repair path in the current exporter; the practical options are `\<model>_int8.tflite`, full-integer `per-channel`, or an experimental custom conversion/calibration flow outside Ultralytics, and if you can reproduce a `per-tensor` path that preserves metrics, updating the existing PR with a minimal test against the [TensorFlow export pipeline](https://docs.ultralytics.com/reference/utils/export/tensorflow) would be helpful.

**Comment by kfrohwergmh** [comment]
Do you got any experience if the per-tensor quantized model is broken for all yolo versions, i.e. yolo5 - yolo26? Or is it a "new" phenomena for yolo11 and yolo26? My question stems from the fact that I've seen that it's possible to switch the quantization type in ultralytics/yolo5.

**Comment by glenn-jocher** [comment]
From what we’ve seen, this is not specific to YOLO11 or YOLO26; it is more a limitation of full-integer `per-tensor` TFLite quantization on detection heads and box decoding, so older YOLO variants can show the same behavior, while classification models tend to tolerate it better. `yolov5` exposed more low-level export knobs because it used a different export stack, but that does not mean `per-tensor` full-integer detection was generally reliable there either; if you have a YOLOv5 repro that keeps metrics close to baseline, please add the exact model, dataset, and export settings to the PR so we can compare pipelines.

**Comment by kfrohwergmh** [comment]
Unfortunately, I don't have an example; I've just seen that `--per-tensor` can be a CLI argument during export.

**Comment by glenn-jocher** [comment]
Got it — in that case I’d treat the `yolov5` `--per-tensor` flag as an older export knob rather than evidence that full-integer `per-tensor` detection was reliable in practice; the current [TFLite export guide](https://docs.ultralytics.com/integrations/tflite) and [exporter reference](https://docs.ultralytics.com/reference/engine/exporter) only document the supported `int8`/`half` paths, so if your hardware strictly requires full-integer `per-tensor`, that still falls into custom conversion and target-specific validation outside the supported Ultralytics export flow today.

**Comment by kfrohwergmh** [comment]
Alright, thank you very much!

**Comment by glenn-jocher** [comment]
You're welcome — if you later get a minimal `per-tensor` full-integer repro that keeps detection metrics close to baseline, please share it here or open a new PR and we can take another look.
